src/mpmc_queue.ml

```diff
@@ -0,0 +1,245 @@
+  let mask = size - 1 in
+  let head = Atomic.make 0 in
+  let tail = Atomic.make 0 in
+  { array; head; tail; mask }
+
+(* [spin_threshold] Number of times on spin on a slot before trying an exit strategy. *)
+let spin_threshold = 30
+
+(* [try_other_exit_every_n] There is two strategies that push/pop can take to fix state (
+   to be able to return without completion). Generally, we want to try to do "rollback" more
+   than "push forward", as the latter adds contention to the side that might already not
+   be keeping up. *)
+let try_other_exit_every_n = 10
+let time_to_try_push_forward n = n mod try_other_exit_every_n == 0
+
+(* [ccas] A slightly nicer CAS. Tries without taking microarch lock first. Use on indices. *)
+let ccas cell seen v =
+  if Atomic.get cell != seen then false else Atomic.compare_and_set cell seen v
+
+let push { array; tail; head; mask; _ } item =
+  let tail_val = Atomic.fetch_and_add tail 1 in
+  let index = tail_val land mask in
+  let cell = Array.get array index in
+
+  (* spin for a bit *)
+  let i = ref 0 in
+  while
+    !i < spin_threshold && not (Atomic.compare_and_set cell None (Some item))
+  do
+    i := !i + 1
+  done;
+
+  (* define clean up function *)
+  let rec take_or_rollback nth_attempt =
+    if Atomic.compare_and_set cell None (Some item) then
+      (* succedded to push *)
+      true
+    else if ccas tail (tail_val + 1) tail_val then (* rolled back tail *)
+      false
+    else if
+      time_to_try_push_forward nth_attempt && ccas head tail_val (tail_val + 1)
+    then (* pushed forward head *)
+      false
+    else (* retry *)
+      take_or_rollback (nth_attempt + 1)
+  in
+
+  (* if succeeded return true otherwise clean up *)
+  if !i < spin_threshold then true else take_or_rollback 0
+
+let take_item cell =
+  let value = Atomic.get cell in
+  if Option.is_some value && Atomic.compare_and_set cell value None then value
+  else None
+
+let pop queue =
+  let ({ array; head; tail; mask; _ } : 'a t) = queue in
+  let head_value = Atomic.get head in
+  let tail_value = Atomic.get tail in
+  if head_value - tail_value >= 0 then None
+  else
+    let old_head = Atomic.fetch_and_add head 1 in
+    let cell = Array.get array (old_head land mask) in
+
+    (* spin for a bit *)
+    let i = ref 0 in
+    let item = ref None in
+    while !i < spin_threshold && not (Option.is_some !item) do
+      item := take_item cell;
+      i := !i + 1
+    done;
+
+    (* define clean up function *)
+    let rec take_or_rollback nth_attempt =
+      let value = Atomic.get cell in
+      if Option.is_some value && Atomic.compare_and_set cell value None then
+        (* dequeued an item, return it *)
+        value
+      else if ccas head (old_head + 1) old_head then (* rolled back head *)
+        None
+      else if
+        time_to_try_push_forward nth_attempt && ccas tail old_head (old_head + 1)
+      then (* pushed tail forward *)
+        None
+      else take_or_rollback (nth_attempt + 1)
+    in
```

Comment:
Same comment as for `push`. If a domain holds a _middle_ value for `old_head`, meaning `old_head` is neither equal to `tail` nor to `head-1` and the queue is empty, none of the 3 possible CAS of this function can work. The only way for the domain to progress is for _other_ domains to change the value of either `head` or `tail`. This is thus not obstruction-free.

It is possible for a domain to hold such a _middle_ value for `old_head` if 3 domains tries to pop simultaneously from an empty queue.